pkg/segment/metadata/segmentmicroindex.go

```diff
@@ -287,14 +306,148 @@ func (sm *SegmentMicroIndex) GetColumns() map[string]bool {
+func (sm *SegmentMicroIndex) getAllColumnsRecSize() map[string]uint32 {
 	retVal := make(map[string]uint32, len(sm.ColumnNames))
 	for colName, colSizeInfo := range sm.ColumnNames {
 		retVal[colName] = colSizeInfo.ConsistentCvalSize
 	}
 	return retVal
 }
 
-func (sm *SegmentMicroIndex) GetRecordCount() uint32 {
+func (sm *SegmentMicroIndex) getRecordCount() uint32 {
 	return uint32(sm.SegMeta.RecordCount)
 }
+
+func GetLoadSsm(segkey string, qid uint64) (*SegmentMicroIndex, int64, error) {
+
+	smi, exists := getMicroIndex(segkey)
+	if !exists {
+		log.Errorf("qid=%d, Seg  %+v does not exist in block meta, but existed in time filtering", qid, segkey)
+		return nil, 0, fmt.Errorf("seg file %+v does not exist in block meta, but existed in time filtering", segkey)
+	}
+
+	totalRequestedMemory := int64(0)
+	if !smi.loadedSearchMetadata {
+		currSearchMetaSize := int64(smi.SearchMetadataSize)
+		totalRequestedMemory += currSearchMetaSize
+		err := GlobalBlockMicroIndexCheckLimiter.TryAcquireWithBackoff(currSearchMetaSize, 10, segkey)
+		if err != nil {
+			log.Errorf("qid=%d, Failed to acquire memory from global pool for search! Error: %v", qid, err)
+			return nil, 0, fmt.Errorf("failed to acquire memory from global pool for search! Error: %v", err)
+		}
+		_, err = smi.loadSearchMetadata([]byte{})
+		if err != nil {
+			log.Errorf("qid=%d, Failed to load search metadata for segKey %+v! Error: %v", qid, smi.SegmentKey, err)
+			return nil, 0, fmt.Errorf("failed to acquire memory from global pool for search! Error: %v", err)
+		}
+	}
+
+	return smi, totalRequestedMemory, nil
+}
+
+func (smi *SegmentMicroIndex) LoadCmiForSearchTime(segkey string,
+	timeFilteredBlocks map[uint16]map[string]bool,
+	colsToCheck map[string]bool, wildcardCol bool,
+	qid uint64) (bool, error) {
+
+	smi.smilock.Lock()
+	defer smi.smilock.Unlock()
+
+	err := GlobalBlockMicroIndexCheckLimiter.TryAcquireWithBackoff(int64(smi.MicroIndexSize),
+		10, segkey)
+	if err != nil {
+		log.Errorf("qid=%d, Failed to acquire memory from global pool for search! err: %v",
+			qid, err)
+		return false, fmt.Errorf("failed to acquire memory from global pool for search! err: %v",
+			err)
+	}
+
+	var allCols map[string]bool
+	if wildcardCol {
+		allCols = smi.GetColumns()
+	} else {
+		allCols = colsToCheck
+	}
+
+	var missingBlockCMI bool
+	blkCmis, err := readCmis(timeFilteredBlocks, false, allCols, smi.VirtualTableName,
+		smi.SegmentKey)
+	if err != nil {
+		log.Errorf("qid=%d, Failed to load cmi for blocks and columns. Num blocks %+v, Num columns %+v. Error: %+v",
+			qid, len(timeFilteredBlocks), len(colsToCheck), err)
+		missingBlockCMI = true
+	} else {
+		smi.setBlockCmis(blkCmis)
+	}
+
+	return missingBlockCMI, nil
+}
+
+func (smi *SegmentMicroIndex) ClearSearchTimeData() {
+
+	if !smi.AreMicroIndicesLoaded() {
+		smi.clearMicroIndices()
+	}
+	if !smi.isSearchMetadataLoaded() {
+		smi.clearSearchMetadata()
+	}
+}
```

Comment:
I might be missing something; this looks like a no-op. e.g., it only clears micro indices if there's no micro indices loaded

Reply:
this function is called during search time, during search we need micro indices for all segs that we are searching for. But we need to then immediately clear them out to avoid mem overload. The rebalance layer only keeps microindices that it is allowed to keep based on the mem percentages. And it then sets the flag isLoaded to true. 
however during search time, we merely load the microindices but do not set the flag, so we are kind of overloading the use of that variable so that we can clear out this microindices. But we also want to mindful, that while we are clearing out and if the rebalance thread kicks in and it loads the micro for this segment, then we dont want to clear it.